data-access/Helpers/Exceptions.cs

```diff
@@ -1,38 +1,151 @@
+using Newtonsoft.Json;
+using SQE.SqeHttpApi.DataAccess.Models;
 
 namespace SQE.SqeHttpApi.DataAccess.Helpers
 {
-    // TODO: Decide whether the controller is aware of these exceptions, or the service translates them
-    public abstract class RepositoryException: Exception
+    // Exceptions of this class will be caught by the middleware and thrown back to HTTP requests with the proper
+    // response codes.  This enables a RESTful experience regardless of whether a request is made via HTTP or by
+    // websocket.  In either case the HTTP router or the Hub will catch a child of the HttpException class
+    // and return that information to the user.
+    // Errors should be thrown properly at the lowest levels of code, this will reduce boiler-plate in higher
+    // level functions like the Services and ensure more consistent responses.
+    public class HttpException : Exception
     {
-        protected RepositoryException(string msg) : base(msg) { }
-    }
+        private readonly int httpStatusCode;
 
-    public class NoPermissionException: RepositoryException
-    {
-        public NoPermissionException(uint? userId, string operation, string entity, uint? entityId) : 
-            base($"User ${userId?.ToString() ?? "anonymous"} can't perform ${operation} on ${entity} ${entityId?.ToString()}")
-        { }
-    }
-    
-    public class ImproperRequestException: RepositoryException
-    {
-        public ImproperRequestException(string operation, string requestError) : 
-            base($"Request ${operation} failed; ${requestError}.")
-        { }
+        public HttpException(HttpStatusCode httpStatusCode, HttpExceptionMessage msg) : base(JsonConvert.SerializeObject(msg))
+        {
+            this.httpStatusCode = (int)httpStatusCode;
+        }
+        
+        public HttpException(HttpStatusCode httpStatusCode, uint internalErrorCode, string msg = null) 
+            : base(JsonConvert.SerializeObject(new HttpExceptionMessage(internalErrorCode, msg)))
+        {
+            this.httpStatusCode = (int)httpStatusCode;
+        }
+
+        public int StatusCode { get { return this.httpStatusCode; } }
     }
-    
-    public class DbFailedWrite: RepositoryException
+
+    public class HttpExceptionMessage
     {
-        public DbFailedWrite() : 
-            base($"Failed writing to DB.")
-        { }
+        public uint internalErrorCode { get; set; }
+        public string message { get; set; }
+        
+        public HttpExceptionMessage(uint internalErrorCode, string message = null)
+        {
+            this.internalErrorCode = internalErrorCode;
+            this.message = message;
+        }
+
     }
     
-    public class DbDetailedFailedWrite: RepositoryException
+    // This is a collection of ready-made errors that can be thrown. They include a standard HTTP status error code
+    // and an internal project error code with accompanying message in English.
+    public static class StandardErrors
     {
```

Comment:
I like the StandardErrors idea. I would probably use specific exception classes for each exception, so I can catch them. So - `HttpException` ( which needs a better name), `PermissionException` which is also abstract, and then `NoPermissionsException`, `NoReadPermissionsException` and so forth. Then `DataException` and all the exceptions that inherit from it.

That way you can catch all `PermissionException` errors in one swoop, and turn all of them in 403 return codes in an HTTP middleware, or turn them into something else with SignalR, and we decouple the logical problems from the HTTP result codes.